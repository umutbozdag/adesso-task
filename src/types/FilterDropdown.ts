export interface FilterDropdownOption {
  label: string;
  name: string;
}

export enum FilterDropdownEmitEnum {
  OPTION_SELECTED = 'option-selected'
}
