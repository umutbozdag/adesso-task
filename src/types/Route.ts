export enum RoutesNameEnum {
  Home = 'Home',
  Units = 'Units',
  UnitDetails = 'UnitDetails',
  NotFound = 'NotFound'
}

export enum RoutesPathEnum {
  Home = '/',
  Units = '/units',
  UnitDetails = '/units/:id',
  NotFound = '/:pathMatch(.*)*'
}

export interface RouteConfigItem {
  path: RoutesPathEnum;
  name: RoutesNameEnum;
  showInNavbar?: boolean;
  label?: string;
}
