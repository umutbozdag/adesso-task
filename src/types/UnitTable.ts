export interface IUnitTableHeaderItem {
  id: string;
  label: string;
}
