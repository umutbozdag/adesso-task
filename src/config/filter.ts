import { UnitAgeEnum } from '@/types/Unit';
import { FilterDropdownOption } from '@/types/FilterDropdown';

export const filterOptionsConfig: FilterDropdownOption[] = [
  {
    label: UnitAgeEnum.All,
    name: UnitAgeEnum.All
  },
  {
    label: UnitAgeEnum.Dark,
    name: UnitAgeEnum.Dark
  },
  {
    label: UnitAgeEnum.Feudal,
    name: UnitAgeEnum.Feudal
  },
  {
    label: UnitAgeEnum.Castle,
    name: UnitAgeEnum.Castle
  },
  {
    label: UnitAgeEnum.Imperial,
    name: UnitAgeEnum.Imperial
  }
];
