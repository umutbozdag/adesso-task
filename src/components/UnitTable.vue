<template>
  <div class="unit-table relative py-8">
    <div class="unit-table__wrapper overflow-y-auto">
      <table class="text-sm text-left text-gray-500 w-full table-auto">
        <thead class="text-xs uppercase bg-gray-900 text-gray-200 sticky top-0">
          <tr>
            <unit-table-header-item
              v-for="tableHeader in tableHeaders"
              :key="tableHeader.id"
              :table-header="tableHeader"
            />
          </tr>
        </thead>
        <tbody class="unit-table__rows h-96">
          <unit-table-row-item
            v-for="(unit, idx) in units"
            :key="unit.id"
            :row-index="idx"
            :unit-detail="unit"
            @row-click="handleRowClick"
          />
        </tbody>
      </table>
    </div>
  </div>
</template>

<script setup lang="ts">
import { IUnitDTO } from '@/types/dto/UnitDTO';
import UnitTableRowItem from './UnitTableRowItem.vue';
import UnitTableHeaderItem from './UnitTableHeaderItem.vue';
import { useRouter } from 'vue-router';
import { RoutesNameEnum } from '@/types/Route';
import { IUnitTableHeaderItem } from '@/types/UnitTable';

interface IUnitTableProps {
  units: IUnitDTO[];
  tableHeaders: IUnitTableHeaderItem[];
}

defineProps<IUnitTableProps>();

const router = useRouter();

const handleRowClick = (item: IUnitDTO) => {
  router.push({
    name: RoutesNameEnum.UnitDetails,
    params: {
      id: item.id
    }
  });
};
</script>

<style lang="scss" scoped>
.unit-table {
  &__wrapper {
    max-height: $table-max-height;
  }
}
</style>
