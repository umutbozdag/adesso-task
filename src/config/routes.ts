import { RouteConfigItem, RoutesNameEnum, RoutesPathEnum } from '@/types/Route';

const routes: Record<RoutesNameEnum, RouteConfigItem> = {
  [RoutesNameEnum.Home]: {
    path: RoutesPathEnum.Home,
    name: RoutesNameEnum.Home,
    showInNavbar: true,
    label: 'Home'
  },
  [RoutesNameEnum.Units]: {
    path: RoutesPathEnum.Units,
    name: RoutesNameEnum.Units,
    showInNavbar: true,
    label: 'Units'
  },
  [RoutesNameEnum.UnitDetails]: {
    path: RoutesPathEnum.UnitDetails,
    name: RoutesNameEnum.UnitDetails
  },
  [RoutesNameEnum.NotFound]: {
    path: RoutesPathEnum.NotFound,
    name: RoutesNameEnum.NotFound
  }
};

export default routes;
