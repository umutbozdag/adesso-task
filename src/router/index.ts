import { RouteRecordRaw, createRouter, createWebHistory } from 'vue-router';
import HomeView from '@/views/HomeView.vue';
import routeConfigs from '@/config/routes';
import { RoutesNameEnum } from '@/types/Route';

export const routes: RouteRecordRaw[] = [
  {
    path: routeConfigs[RoutesNameEnum.Home].path,
    name: routeConfigs[RoutesNameEnum.Home].name,
    component: HomeView
  },
  {
    path: routeConfigs[RoutesNameEnum.Units].path,
    name: routeConfigs[RoutesNameEnum.Units].name,
    component: () => import(/* webpackChunkName: "Units" */ '@/views/UnitsView.vue')
  },
  {
    path: routeConfigs[RoutesNameEnum.UnitDetails].path,
    name: routeConfigs[RoutesNameEnum.UnitDetails].name,
    component: () => import(/* webpackChunkName: "UnitDetails" */ '@/views/UnitDetailsView.vue')
  },
  {
    path: routeConfigs[RoutesNameEnum.NotFound].path,
    component: () => import(/* webpackChunkName: "NotFound" */ '@/components/NotFound.vue')
  }
];

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes
});

export default router;
