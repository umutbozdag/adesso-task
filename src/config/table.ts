import { IUnitTableHeaderItem } from '@/types/UnitTable';

export function createTableHeaders(): IUnitTableHeaderItem[] {
  return [
    {
      id: 'id',
      label: 'ID'
    },
    {
      id: 'name',
      label: 'Name'
    },
    {
      id: 'age',
      label: 'Age'
    },
    {
      id: 'costs',
      label: 'Costs'
    }
  ];
}
