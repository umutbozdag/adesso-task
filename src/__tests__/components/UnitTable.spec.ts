import UnitTable from '@/components/UnitTable.vue';
import { VueWrapper, mount } from '@vue/test-utils';
import { expect, it, describe, beforeAll, vi } from 'vitest';
import { ComponentPublicInstance } from 'vue';
import { generateArrayOfUnits } from '../mocks/unit.mocks';
import router from '@/router';
import { RoutesNameEnum } from '@/types/Route';
import { createTableHeaders } from '@/config/table';

describe('UnitTable', () => {
  let wrapper: VueWrapper<ComponentPublicInstance & { [key: string]: any }>;

  beforeAll(() => {
    wrapper = mount(UnitTable, {
      global: {
        plugins: [router]
      },
      props: {
        tableHeaders: createTableHeaders(),
        units: generateArrayOfUnits(1)
      }
    });
  });

  it('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('handleRowClick pushing to Unit Details page', async () => {
    const routerPushSpy = vi.spyOn(router, 'push');
    const unitMock = generateArrayOfUnits(1)[0];
    wrapper.vm.handleRowClick(unitMock);
    expect(routerPushSpy).toHaveBeenCalledWith({
      name: RoutesNameEnum.UnitDetails,
      params: {
        id: unitMock.id
      }
    });
  });
});
