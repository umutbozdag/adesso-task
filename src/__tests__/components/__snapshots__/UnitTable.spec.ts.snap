// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UnitTable > renders correctly 1`] = `
"<div data-v-b7cca949=\\"\\" class=\\"unit-table relative py-8\\">
  <div data-v-b7cca949=\\"\\" class=\\"unit-table__wrapper overflow-y-auto\\">
    <table data-v-b7cca949=\\"\\" class=\\"text-sm text-left text-gray-500 w-full table-auto\\">
      <thead data-v-b7cca949=\\"\\" class=\\"text-xs uppercase bg-gray-900 text-gray-200 sticky top-0\\">
        <tr data-v-b7cca949=\\"\\">
          <th data-v-b7cca949=\\"\\" class=\\"px-6 py-3\\">ID</th>
          <th data-v-b7cca949=\\"\\" class=\\"px-6 py-3\\">Name</th>
          <th data-v-b7cca949=\\"\\" class=\\"px-6 py-3\\">Age</th>
          <th data-v-b7cca949=\\"\\" class=\\"px-6 py-3\\">Costs</th>
        </tr>
      </thead>
      <tbody data-v-b7cca949=\\"\\" class=\\"unit-table__rows h-96\\">
        <tr data-v-b7cca949=\\"\\" class=\\"border-b bg-gray-800 border-gray-700 text-gray-300 cursor-pointer hover:bg-gray-700 table__row--even\\">
          <!-- TODO: get 'id' from header and render based on that -->
          <!-- OR -->
          <!-- TODO: use 'label' props to pass data and use a slot for more advanced things (rendering costs array) -->
          <td scope=\\"row\\" class=\\"px-6 py-4 font-medium whitespace-nowrap\\">1</td>
          <td class=\\"px-6 py-4\\">Archer</td>
          <td class=\\"px-6 py-4\\">Dark</td>
          <td class=\\"px-6 py-4\\"><span>Wood: 25, </span><span>Gold: 45<!--v-if--></span></td>
        </tr>
      </tbody>
    </table>
  </div>
</div>"
`;
